Sources/NIOSOCKS/Channel Handlers/SOCKSClientHandler.swift

```diff
@@ -0,0 +1,118 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// Data was unexpectedly written or read before the SOCKS proxy
+/// connection has been fully established.
+public struct ProxyNotEstablished: Error {
+    
+}
+
+/// Connects to a SOCKS server to establish a proxied connection
+/// to a host. This handler should be inserted at the beginning of a
+/// channel's pipeline.
+public class SOCKSClientHandler: ChannelDuplexHandler {
+    
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    public typealias OutboundIn = ByteBuffer
+    public typealias OutboundOut = ByteBuffer
+    
+    public let supportedAuthenticationMethods: [AuthenticationMethod]
+    public let targetAddress: AddressType
+    public let targetPort: UInt16
+    
+    private var state: ClientStateMachine
+    private var buffered: ByteBuffer
+    
+    public init(supportedAuthenticationMethods: [AuthenticationMethod], targetAddress: AddressType, targetPort: UInt16) {
+        precondition(supportedAuthenticationMethods.count <= 255)
```

Comment:
Do we maybe want to add an error message, why this must be less than or equal 255?